File: apps/api/src/modules/conversations/sender.ts
// Adapter pengiriman WhatsApp: official (Meta Cloud API) & unofficial
// (gateway Baileys via HTTP). Dipilih per channel saat kirim.
export interface SendResult {
  ok: boolean;
  externalId?: string;
  error?: string;
}

interface ChannelConfig {
  type?: string;
  phone_number_id?: string;
  access_token?: string;
  gateway_url?: string;
  api_key?: string;
  session?: string;
}

async function sendOfficial(to: string, body: string, cfg: ChannelConfig): Promise<SendResult> {
  if (!cfg.phone_number_id || !cfg.access_token) {
    return { ok: false, error: "Config official belum lengkap (phone_number_id/access_token)" };
  }
  try {
    const res = await fetch(`https://graph.facebook.com/v21.0/${cfg.phone_number_id}/messages`, {
      method: "POST",
      headers: { Authorization: `Bearer ${cfg.access_token}`, "Content-Type": "application/json" },
      body: JSON.stringify({ messaging_product: "whatsapp", to, type: "text", text: { body } }),
    });
    const data = (await res.json().catch(() => ({}))) as { messages?: { id: string }[]; error?: { message: string } };
    if (!res.ok) return { ok: false, error: data.error?.message ?? `HTTP ${res.status}` };
    return { ok: true, externalId: data.messages?.[0]?.id };
  } catch (e) {
    return { ok: false, error: `Gateway tidak terjangkau: ${(e as Error).message}` };
  }
}

async function sendUnofficial(to: string, body: string, cfg: ChannelConfig): Promise<SendResult> {
  if (!cfg.gateway_url) {
    return { ok: false, error: "Config unofficial belum lengkap (gateway_url)" };
  }
  try {
    const res = await fetch(`${cfg.gateway_url.replace(/\/$/, "")}/send`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...(cfg.api_key ? { "X-Api-Key": cfg.api_key } : {}) },
      body: JSON.stringify({ to, message: body, session: cfg.session ?? "default" }),
    });
    const data = (await res.json().catch(() => ({}))) as { id?: string; error?: string };
    if (!res.ok) return { ok: false, error: data.error ?? `HTTP ${res.status}` };
    return { ok: true, externalId: data.id };
  } catch (e) {
    return { ok: false, error: `Gateway tidak terjangkau: ${(e as Error).message}` };
  }
}

export async function sendWhatsApp(
  to: string,
  body: string,
  channel: { type: string; config: unknown },
): Promise<SendResult> {
  const cfg = (channel.config ?? {}) as ChannelConfig;
  if (channel.type === "official") return sendOfficial(to, body, cfg);
  return sendUnofficial(to, body, cfg);
}

File: internal/wa/sender.go
package wa

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Sender mengirim pesan via channel official (Meta Cloud API)
// atau unofficial (gateway Baileys via HTTP).
type Result struct {
	OK         bool
	ExternalID string
	Err        string
}

type sender struct{ client *http.Client }

// Sender pengirim WA (official/unofficial), bisa di-mock di test.
type Sender interface {
	Send(channelType string, cfg map[string]string, to, body string) Result
}

func New() Sender { return &sender{client: &http.Client{Timeout: 20 * time.Second}} }

// cfg: official {phone_number_id, access_token}; unofficial {gateway_url, api_key, session}
func (s *sender) Send(channelType string, cfg map[string]string, to, body string) Result {
	if channelType == "official" {
		return s.official(to, body, cfg)
	}
	return s.unofficial(to, body, cfg)
}

func (s *sender) official(to, body string, cfg map[string]string) Result {
	pid, tok := cfg["phone_number_id"], cfg["access_token"]
	if pid == "" || tok == "" {
		return Result{Err: "config official belum lengkap (phone_number_id/access_token)"}
	}
	payload, _ := json.Marshal(map[string]any{
		"messaging_product": "whatsapp", "to": to, "type": "text", "text": map[string]string{"body": body},
	})
	req, _ := http.NewRequest("POST", "https://graph.facebook.com/v21.0/"+pid+"/messages", bytes.NewReader(payload))
	req.Header.Set("Authorization", "Bearer "+tok)
	req.Header.Set("Content-Type", "application/json")
	resp, err := s.client.Do(req)
	if err != nil {
		return Result{Err: "gateway tidak terjangkau: " + err.Error()}
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	var data struct {
		Messages []struct {
			ID string `json:"id"`
		} `json:"messages"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	_ = json.Unmarshal(b, &data)
	if resp.StatusCode >= 300 {
		msg := fmt.Sprintf("HTTP %d", resp.StatusCode)
		if data.Error != nil {
			msg = data.Error.Message
		}
		return Result{Err: msg}
	}
	id := ""
	if len(data.Messages) > 0 {
		id = data.Messages[0].ID
	}
	return Result{OK: true, ExternalID: id}
}

func (s *sender) unofficial(to, body string, cfg map[string]string) Result {
	gw := strings.TrimSuffix(cfg["gateway_url"], "/")
	if gw == "" {
		return Result{Err: "config unofficial belum lengkap (gateway_url)"}
	}
	payload, _ := json.Marshal(map[string]string{"to": to, "message": body, "session": first(cfg["session"], "default")})
	req, _ := http.NewRequest("POST", gw+"/send", bytes.NewReader(payload))
	req.Header.Set("Content-Type", "application/json")
	if cfg["api_key"] != "" {
		req.Header.Set("X-Api-Key", cfg["api_key"])
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return Result{Err: "gateway tidak terjangkau: " + err.Error()}
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	var data map[string]string
	_ = json.Unmarshal(b, &data)
	if resp.StatusCode >= 300 {
		return Result{Err: first(data["error"], fmt.Sprintf("HTTP %d", resp.StatusCode))}
	}
	return Result{OK: true, ExternalID: data["id"]}
}

func first(a, b string) string {
	if a != "" {
		return a
	}
	return b
}
